library/core/src/ptr/const_ptr.rs

```diff
@@ -232,23 +232,26 @@ impl<T: ?Sized> *const T {
     ///
     /// # Safety
     ///
-    /// The resulting pointer does not need to be in bounds, but it is
-    /// potentially hazardous to dereference (which requires `unsafe`).
+    /// This operation itself is always safe, but using the resulting pointer is not.
     ///
-    /// In particular, the resulting pointer remains attached to the same allocated
-    /// object that `self` points to. It may *not* be used to access a
-    /// different allocated object. Note that in Rust,
-    /// every (stack-allocated) variable is considered a separate allocated object.
+    /// The resulting pointer remains attached to the same allocated object that `self` points to.
+    /// It may *not* be used to access a different allocated object. Note that in Rust, every
+    /// (stack-allocated) variable is considered a separate allocated object.
     ///
-    /// In other words, `x.wrapping_offset((y as usize).wrapping_sub(x as usize) / size_of::<T>())`
-    /// is *not* the same as `y`, and dereferencing it is undefined behavior
-    /// unless `x` and `y` point into the same allocated object.
+    /// In other words, `let z = x.wrapping_add((y as usize).wrapping_sub(x as usize) /
+    /// size_of::<T>())` does *not* make `z` the same as `y`: `z` is still attached to the object `x` is
```

Comment:
I’m not 100% happy with the fact that the first "example code" in the safety remark is one that doesn’t even mention `wrapping_offset`. The same applies to the documentation of `wrapping_sub`.

Reply:
I guess we could forcefully cast... but that just obscures the point even more as the use of `usize::wrapping_sub` already does. I'd prefer to just write `x.wrapping_add(y - x)`, since then it is much more clear what is happening.

I did some more editing, what do you think?

Reply:
Yes, clarity is the most important thing here (while staying "correct" is only important to avoid confusion). Stating some preconditions on the size of `T` and overflow, like you did now, is a nice approach.